enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,254 @@
+---
+title: on-prem-configurable-lb-dns-stack
+authors:
+  - "@yboaron"
+reviewers:
+  - "@sttts"
+  - "@deads2k"
+  - "@celebdor"
+  - "@bcrochet"
+  - "@bnemec"
+approvers:
+  - "@"
+creation-date: 2020-02-11
+last-updated: 2020-02-11
+status: 
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# On-prem: configurable DNS/LB self-hosted stack 
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the current implementation,  the self-hosted DNS/LB stack runs under the MCO umbrella by default in all on-prem platforms and there's no option to configure it. 
+Some customers with high production needs usually have their own external Load Balancing and DNS resolution. In addition, there are cases in which some traffic used to provide these self-hosted services is disallowed in the customer's network.                                                                                                                                                                                                                                        The outcome of such scenarios is that cluster's resources are spent on providing unused services and generating traffic against certain customers network design.
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+Allow customers to adjust the self hosted DNS/LB stack according to the needs they really have instead of providing them in a generic way. 
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+
+
+### Non-Goals
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/baremetal) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by both MCO and the new operator.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run Coredns-MDNS, HAProxy and MDNS-publisher self-hosted stack components as a K8S applications (while Keepalived will continue running as static pod)
```

Comment:
how can that work? The apiserver VIP is served by the haproxy.

Reply:
Compare https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md#api-load-balancing

Reply:
HAProxy pod includes a sidecar container responsible to redirect (using iptable rule)  API traffic to the loadbalancer frontend. 
So, if only Keepalived runs we have HA API access, and if both Keepalived and HAProxy are running we have HA, LB API access.

The installer documentation should be updated.

Reply:
I am still not convinced about this plan. When haproxy dies (and it will die early – earlier than kube-apiserver as static pod – during reboot), connections are cut off. This risks the availability guarantees of kube-apiserver.

/hold

until this is resolved.

Reply:
Can you elaborate on which cases concern you? 

Regarding your concerns about master reboot process, this case is relevant in the context of kube-api availability only when the rebooted master holds the VIP. 

If HAProxy runs as DS we will get the next points in time in case master-0 holds VIP being rebooted:

t0 - master-0 HAProxy pod dies and redirect the local API requests back to localhost kube-apiserver

t1 - master-0 Keepalived or kube-apiserver die

t2 - Keepalived on master-1 detects that Keepalived on master-0 failed, assigns the VIP and sends GARP. 

t2 - t1 = From what I recall from previous tests  (I'll double-check that) 
for cold reboot it should be : ~ 4000 mSec
for warm reboot : its  ~ 350 mSeconds  (detection time + GARP )

t1 - t0 =  Do you have a time estimation for this? 


- I assume that t1-t0 should be significantly smaller than t2 - t1  so it should have a tiny impact on the total time.

- If will be needed to consider the t1-t0 time, we can find a way for HAProxy to trigger Keepalived shutdown on local node (both of them run on host networking)
-
- It makes sense to add a goal (and update also tests section)  to OEP for kube-apiserver downtime should be similar to the values ​​with static pods

Reply:
One correction: Keepalived will fail over the VIP as soon as HAProxy goes down. It prefers nodes with working loadbalancers.

So there's two potential disconnections: 1) When HAProxy goes down. While the api will remain available, I doubt connections made through HAProxy will continue working. 2) When the VIP fails over.

There should be no significant outage of the API at all though. In both of those cases, the API will be immediately available via another path. The VIP change is unavoidable in our LB architecture. It's possible we could do something to move the VIP even before HAProxy goes down and avoid that step, however.

Reply:
> One correction: Keepalived will fail over the VIP as soon as HAProxy goes down. It prefers nodes with working loadbalancers.

Yep (10x for the reminder), this should handle the VIP transition in the case of node drain

Reply:
A. Coredns-mdns , mdns-publisher and HAProxy pods should run as DaemonSets (the OEP was updated) , also added a goal (and test respectively) for kube-api downtime.


B. I briefly checked master node drain in this context, and it seems that DaeomonSets pods are not evicted from the node. 
As far as I understand, for node drain use case, we should have similar behavior as with static pods. Is it correct?

Reply:
>  1) When HAProxy goes down. While the api will remain available, I doubt connections made through HAProxy will continue working.
> There should be no significant outage of the API at all though. In both of those cases, the API will be immediately available via another path. 

This breaks existing connection, hence this is an outage of the API we cannot accept. We have a protocol for graceful termination that applies to both kube-apiserver restarts without and with reboot (compare https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md). Connection through haproxy are terminated when haproxy dies. This is not acceptable.

Reply:
> As far as I understand, for node drain use case, we should have similar behavior as with static pods. Is it correct?

@rphillips @sjenning can comment on this I hope. Do DaemonSet pods live as long as static pods during reboot? What about "critical pods"? I remember there is a concept like that in kubelet to keep e.g. kube-apiserver up very long.

Reply:
> This breaks existing connection, hence this is an outage of the API we cannot accept. We have a protocol for graceful termination that applies to both kube-apiserver restarts without and with reboot (compare https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md). Connection through haproxy are terminated when haproxy dies. This is not acceptable.

That's correct, In order not to break the graceful termination process of the kube-api server the HAProxy pod should be 
stopped at the same time (or after) the kube-apiserver static pod.

We should be able to implement this requirement using https://github.com/kubernetes/kubernetes/pull/96129 .

I'll update the OEP.